File: output/StarCoder/codenet/C++/Python/s345215315.py
import sys

h,w = map(int,input().split())
s = [input() for _ in range(h)]
ans = 20

def dfs(x,y,cnt):
    global ans
    if cnt > ans:return
    if x == h-1:
        ans = min(ans,cnt)
        return
    for i in range(8):
        nx = x + dx[i]
        ny = y + dy[i]
        if nx < 0 or nx >= h or ny < 0 or ny >= w:continue
        if s[nx][ny]!= 'X':continue
        c = True
        while c and s[nx][ny] == 'X':
            s[nx][ny] = '.'
            nx += dx[i]
            ny += dy[i]
            if nx < 0 or nx >= h or ny < 0 or ny >= w:
                if nx >= h:ans = min(ans, cnt+1)
                c = False
        if c:dfs(nx,ny,cnt+1)
        while nx!= x or ny!= y:
            nx -= dx[i]
            ny -= dy[i]
            s[nx][ny] = 'X'
        s[x][y] = '.'

sx,sy = -1,-1
for i in range(h):
    for j in range(w):
        if s[i][j] == 'O':
            sx,sy = i,j
            s[i][j] = '.'
            break
dfs(sx,sy,0)
if ans == 20:print(-1)
else:print(ans)

File: output/StarCoder/codenet/C++/Java/s345215315.java
import java.util.*;
import java.io.*;

public class s345215315 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int h = 19,w = 15;
        int dx[] = {0,1,0,-1,1,1,-1,-1};
        int dy[] = {1,0,-1,0,1,-1,1,-1};
        String s[] = new String[19];
        int ans = Integer.MAX_VALUE;
        for(int i = 0;i < h;i++)s[i] = sc.next();
        int sx,sy;
        for(int i = 0;i < h;i++){
            for(int j = 0;j < w;j++){
                if(s[i].charAt(j) == 'O'){
                    sx = i;
                    sy = j;
                    s[i] = s[i].substring(0,j) + "." + s[i].substring(j+1);
                    break;
                }
            }
        }
        dfs(s,sx,sy,0,ans);
        if(ans == Integer.MAX_VALUE)System.out.println(-1);
        else System.out.println(ans);
    }
    public static void dfs(String s[][],int x,int y,int cnt,int ans){
        if(cnt > 20)return;
        if(x == 18){
            ans = Math.min(ans,cnt);
            return;
        }
        for(int i = 0;i < 8;i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            if(nx < 0 || nx >= 19 || ny < 0 || ny >= 15)continue;
            if(s[nx][ny]!= 'X')continue;
            boolean c = true;
            while(c && s[nx][ny] == 'X'){
                s[nx][ny] = '.';
                nx += dx[i];
                ny += dy[i];
                if(nx < 0 || nx >= 19 || ny < 0 || ny >= 15){
                    if(nx >= 19)ans = Math.min(ans, cnt+1);
                    c = false;
                }
            }
            if(c)dfs(s,nx,ny,cnt+1,ans);
            while(nx!= x || ny!= y){
                nx -= dx[i];
                ny -= dy[i];
                s[nx][ny] = 'X';
            }
            s[x][y] = '.';
        }
    }
}